src/sparse_int_set.jl

```diff
@@ -0,0 +1,258 @@
+import Base: @propagate_inbounds, zip
+
+const INT_PER_PAGE = div(ccall(:jl_getpagesize, Clong, ()), sizeof(Int))
+
+#TODO: Batch creation and allocation
+mutable struct SparseIntSet
```

Comment:
Yea, I didn't see the benefit of immutable if it has `Vector`s anyway, is there any? I don't mind making it immutable, just for my own understanding

Reply:
Yeah, it is better, not entirely sure why
we may have the stack-reference-to-heap-allocated object thing working


```
julia> struct Foo
       x::Vector{Int}
       end

julia> mutable struct MFoo
       x::Vector{Int}
       end

julia> function testem(T)
       first([T([1,2,i]) for i in 1:10_000])
       end
testem (generic function with 1 method)
```

Timing on Julia 1.4-DEV
```
julia> @btime testem(Foo)
  1.507 ms (20005 allocations: 1.30 MiB)
Foo([1, 2, 1])

julia> @btime testem(MFoo)
  1.633 ms (20005 allocations: 1.30 MiB)
MFoo([1, 2, 1])
```

Timing on julia 1.0
```
julia> @btime testem(MFoo)
  1.112 ms (29495 allocations: 1.44 MiB)
MFoo([1, 2, 1])

julia> @btime testem(Foo)
  1.209 ms (29495 allocations: 1.44 MiB)
Foo([1, 2, 1])
```

Reply:
So you mean to keep it nutable for now since it's faster on current Julia, or make it immutable to future proof?

Reply:
oops I didn't splot that it flippped.

on current stable 1.2 it is also faster to be mutable
```
julia> @btime testem(MFoo)
  1.437 ms (20005 allocations: 1.30 MiB)
MFoo([1, 2, 1])

julia> @btime testem(Foo)
  1.571 ms (20005 allocations: 1.30 MiB)
Foo([1, 2, 1])
```

On 1.3-rc2
is where it becomes faster to be immutable
```
julia> @btime testem(Foo)
  1.535 ms (20005 allocations: 1.30 MiB)
Foo([1, 2, 1])

julia> @btime testem(MFoo)
  1.720 ms (20005 allocations: 1.30 MiB)
MFoo([1, 2, 1])
```

Lets go immutable, for the future proofing.
Since 1.3 will become the stable release soon

Reply:
Aside from the fact that mutable vs immutable speed switches, it's kind of sad/scary to see that with each next minor release this is getting slower...